src/duk_debugger.c

```diff
@@ -1639,6 +1690,37 @@ DUK_LOCAL void duk__debug_process_message(duk_hthread *thr) {
 	return;
 }
 
+/* Halt execution and enter a debugger message loop until execution is resumed
+ * by the client. PC for the current activation may be temporarily decremented
+ * so that the "current" instruction will be highlighted in the client.
+ */
+
+DUK_INTERNAL void duk_debug_halt_execution(duk_hthread *thr, duk_bool_t use_prev_pc) {
+	duk_activation *act;
+	duk_instr_t *old_pc;
+
+	DUK_HEAP_SET_PAUSED(thr->heap);
+
+	/* Decrement PC if that was requested, this requires a PC sync */
+	act = duk_hthread_get_current_activation(thr);
+	duk_hthread_sync_currpc(thr);
+	old_pc = act->curr_pc;
+	if (use_prev_pc && act->curr_pc > 0) {
+		act->curr_pc--;
+	}
+
+	/* Process debug messages until we are no longer paused */
+	thr->heap->dbg_processing = 1;
+	duk_debug_send_status(thr);
+	while (thr->heap->dbg_paused) {
+		DUK_ASSERT(thr->heap->dbg_processing);
+		duk_debug_process_messages(thr, 0 /* no_block */);
+	}
+	thr->heap->dbg_processing = 0;
+
+	act->curr_pc = old_pc;  /* restore PC */
```

Comment:
I agree, it's just it's not entirely clear what, concretely, can be done.  It would be nice to make the assumption explicit with an assert, but where to put it? If an error is thrown, we lose all information that the PC was ever changed, so it's hard to verify that it was indeed changed back.